### ISSUE TITLE
base64 encode and decode

### ISSUE THREAD

**Issue by devth** [yetibot/yetibot#687]

**Comment by themistoklik** [comment]
Hello,
I'm a beginner in Clojure and I'd like to contribute by working on this. I'm thinking on building it to be called like` !base64 {encode | decode} <string>`. So basically encode and decode will be subcommands. Is this agreeable? 

I've poked around the codebase and what I can't figure out is how the hooking mechanism works. I built a prototype for encoding that works like `!base64 <string> ` by cloning how echo hooks stuff. Can you point me to a resource that explains how to pass command args to handlers?

**Comment by devth** [comment]
Hi, sounds good regarding `!base64 {encode | decode} <string>`!

You can use the [jira command](https://github.com/yetibot/yetibot/blob/master/src/yetibot/commands/jira.clj#L247) as a more complex example that has various sub commands. The result of the regex you use in `cmd-hook` is available in the handler as the `:match` key, e.g. [parse-cmd](https://github.com/yetibot/yetibot/blob/master/src/yetibot/commands/jira.clj#L225).

So for `base64` it'd look something like:

```clojure
(defn encode-cmd
  "base64 encode <string>"
  [{[_ string] :match}]
  ;; ...
  )

(cmd-hook #"base64"
  #"^encode\s(.+)" encode-cmd
  )
```

The above example uses destructuring to extract the 2nd item out of the `:match` vector. Another thing you could do to get more familiar with how args are passed is to print the arguments passed, like:

```clojure
(defn encode-cmd
  [opts]
  (prn "encode-cmd" opts)
  )
```

Let me know if any of that doesn't make sense or ping me on slack!

**Comment by themistoklik** [comment]
Hello,

please check out [devth#688](https://github.com/yetibot/yetibot/pull/688) and let me know for any fixes.

**Comment by devth** [comment]
Fixed by #688!
